lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2013,17 +2002,20 @@ static void checkAccessibility(TypeChecker &TC, const Decl *D) {
       }
     });
 
-    if (minAccessScope) {
-      auto minAccess = accessibilityFromScopeForDiagnostics(minAccessScope);
+    if (!minAccessScope.isPublic()) {
+      auto minAccess = minAccessScope.accessibilityForDiagnostics();
       bool isExplicit =
         SD->getAttrs().hasAttribute<AccessibilityAttr>() ||
         SD->getDeclContext()->getAsProtocolOrProtocolExtensionContext();
       auto diagID = diag::subscript_type_access;
       if (downgradeToWarning == DowngradeToWarning::Yes)
         diagID = diag::subscript_type_access_warn;
+      auto subscriptDeclAccess = isExplicit
+        ? SD->getFormalAccess()
+        : minAccessScope.requiredAccessibilityForDiagnostics();
```

Comment:
In order to make solve this situation.

```diff
+subscript (a: PrivateStruct) -> Int { return 0 } // expected-error {{subscript must be declared fileprivate because its index uses a private type}}
-subscript (a: PrivateStruct) -> Int { return 0 } // expected-error {{subscript must be declared fileprivate because its index uses a fileprivate type}}
+subscript (a: Int) -> PrivateStruct { return PrivateStruct() } // expected-error {{subscript must be declared fileprivate because its element type uses a private type}}
-subscript (a: Int) -> PrivateStruct { return PrivateStruct() } // expected-error {{subscript must be declared fileprivate because its element type uses a fileprivate type}}
```

If the accessibility declaration isn't explicit, `minAccessScope` was used as the required access scope for subscript declaration, which leads to this *incorrect* message - `{{subscript must be declared private because its index uses a private type}}`.

Doing this change I tweaked the parameters for the `subscript_type_access` diagnostics in such way, so it complained on the inferred accessibility level in explicit case, and suggest required in implicit case:
`%select{must be declared %select{...}1 | cannot be declared %select{...}1}0` where `1 parameter` depends on `0 parameter`

Am I missing something here?